R/sheets_speedread.R

```diff
@@ -0,0 +1,221 @@
+#' Read Sheet as CSV
+#'
+#' @description
+#' This function uses a quick-and-dirty method to read a Sheet that bypasses the
+#' Sheets API entirely. It can be much faster than [sheets_read()] -- noticeably
+#' so for "large" spreadsheets. `sheets_speedread()` forms a special URL that
+#' reads Sheet values via CSV, then sends that data through vroom or readr.
+#' Here are the limitations we must accept to get faster reading:
+#' * Only cell values are available, not formats.
+#' * We can't target a named range as the `range`.
+#' * We have no access to the data type of a cell, i.e. we don't know that it's
+#'   logical, numeric, or datetime. That must be re-discovered based on the
+#'   CSV data (or specified by the user).
+#' * Auth and error handling have to be handled a bit differently internally,
+#'   which may lead to behaviour that differs from other functions in
+#'   googlesheets4.
+#'
+#' @section The export URL:
+#'
+#' The export URL supports various query parameters. We always send
+#' `format=csv` and, optionally, `gid=SHEET_ID` and/or `range=A2:D5` (`range`
+#' must really be cell range in this case, not a sheet name or named range):
+#'
+#' ```
+#' # here is the base URL:
+#' https://docs.google.com/spreadsheets/d/SPREADSHEET_ID/export
+#'
+#' # examples of the query part
+#' ?format=csv
+#' ?format=csv?range=CELL_RANGE
+#' ?format=csv?gid=SHEET_ID
+#' ?format=csv?range=A1_range?gid=SHEET_ID
+#' ```
+#'
+#' The export URL supports the use of a bearer token in the Authorization
+#' header.
+#'
+#' Gist of various query parameters that work for the export URL, with special
+#' emphasis on the PDF format, which doesn't help us:
+#'   * <https://gist.github.com/Spencer-Easton/78f9867a691e549c9c70>
+#'
+#' @section The datasource URL:
+#'
+#' The "Chart Tools datasource protocol" is another way to access values in
+#' Sheets, designed to make Sheets play nicely with the Google Charts API. We
+#' can use the datasource URL as another way to get data as CSV (or JSON,
+#' etc.). It is more powerful than the export URL, in that it supports an
+#' entire SQL-like query language.
```

Comment:
This URL is of great long-term interest to us because it will likely play a role in some of the dplyr-ish stuff.

Reply:
Maaaaaybe; I was thinking the dplyr backend would actually modify a sheet so you could see the data changing.